src/main/cpp/profiler/ProfilerJni.cpp

```diff
@@ -0,0 +1,530 @@
+  uint8_t* data_;
+  size_t size_;
+  size_t valid_size_;
+};
+
+// Queue of profile buffers that have been filled with profile data.
+struct completed_buffer_queue {
+  // Gets the next available buffer of profile data, blocking until a buffer is available
+  // or the queue is shutdown. If the queue is shutdown, a nullptr is returned.
+  std::unique_ptr<profile_buffer> get()
+  {
+    std::unique_lock lock(lock_);
+    cv_.wait(lock, [this] { return shutdown_ || buffers_.size() > 0; });
+    if (buffers_.size() > 0) {
+      auto result = std::move(buffers_.front());
+      buffers_.pop();
+      return result;
+    }
+    return std::unique_ptr<profile_buffer>(nullptr);
+  }
+
+  void put(std::unique_ptr<profile_buffer>&& buffer)
+  {
+    std::unique_lock lock(lock_);
+    if (!shutdown_) {
+      buffers_.push(std::move(buffer));
+      lock.unlock();
+      cv_.notify_one();
+    }
+  }
+
+  void shutdown()
+  {
+    std::unique_lock lock(lock_);
+    shutdown_ = true;
+    lock.unlock();
+    cv_.notify_one();
+  }
+
+ private:
+  std::mutex lock_;
+  std::condition_variable cv_;
+  std::queue<std::unique_ptr<profile_buffer>> buffers_;
+  bool shutdown_ = false;
+};
+
+// Stack of profile buffers that are ready to be filled with profile data.
+struct free_buffer_tracker {
+  explicit free_buffer_tracker(size_t size) : buffer_size_(size) {}
+
+  // Returns the next available profile buffer or creates one if none are available.
+  std::unique_ptr<profile_buffer> get()
+  {
+    {
+      std::lock_guard lock(lock_);
+      if (buffers_.size() > 0) {
+        auto result = std::move(buffers_.top());
+        buffers_.pop();
+        return result;
+      }
+    }
+    return std::make_unique<profile_buffer>(buffer_size_);
+  }
+
+  void put(std::unique_ptr<profile_buffer>&& buffer)
+  {
+    buffer->set_valid_size(0);
+    std::lock_guard lock(lock_);
+    if (buffers_.size() < NUM_CACHED_BUFFERS) {
+      buffers_.push(std::move(buffer));
+    } else {
+      buffer.reset(nullptr);
+    }
+  }
+
+ private:
+  static constexpr size_t NUM_CACHED_BUFFERS = 2;
+  std::mutex lock_;
+  std::stack<std::unique_ptr<profile_buffer>> buffers_;
+  size_t buffer_size_;
+};
+
+void writer_thread_process(JavaVM* vm,
+                           jobject j_writer,
+                           size_t buffer_size,
+                           size_t flush_threshold);
+
+struct subscriber_state {
+  CUpti_SubscriberHandle subscriber_handle;
+  jobject j_writer;
+  std::thread writer_thread;
+  free_buffer_tracker free_buffers;
+  completed_buffer_queue completed_buffers;
+  bool has_cupti_callback_errored;
+  bool is_shutdown;
+
+  subscriber_state(jobject writer, size_t buffer_size)
+    : j_writer(writer),
+      free_buffers(buffer_size),
+      has_cupti_callback_errored(false),
+      is_shutdown(false)
+  {
+  }
+};
+
+// Global variables
+subscriber_state* State = nullptr;
+uint64_t Flush_period_msec;
+uint64_t Last_flush_time_msec;
```

Comment:
Do we need some kind of locking for this? if so it would be good to have a comment explaining that here. The flush period should be mostly static once it is set.

Reply:
We don't need any locking for Flush_period_msec.  It's set once on startup before any callbacks try to examine it.  The more interesting one is Last_flush_time_msec which theoretically could be set by multiple threads.  This probably ends up being set atomically in practice, but probably prudent to make it an atomic to guarantee it.  We shouldn't care which thread "wins" a race for this, as they're all trying to set it to the current time.  Worst-case, we end up flushing ever so slightly more often than we should because a slightly stale timestamp overwrites a newer one.